Source/WebKit/UIProcess/WebAuthentication/Cocoa/CcidConnection.mm

```diff
@@ -74,47 +74,55 @@ inline bool compareCcidVersion(NSData *data, const uint8_t version[], size_t ver
+        ASSERT(RunLoop::isMain());
+        if (!weakThis)
+            return;
+        if (response == Vector { kCtapNfcAppletSelectionU2f, sizeof(kCtapNfcAppletSelectionU2f) }
+            || response == Vector { kCtapNfcAppletSelectionCtap, sizeof(kCtapNfcAppletSelectionCtap) }) {
+            if (weakThis && weakThis->m_service)
+                weakThis->m_service->didConnectTag();
             return;
         }
-            [m_smartCard transmitRequest:adoptNS([[NSData alloc] initWithBytes:kCtapNfcU2fVersionCommand length:sizeof(kCtapNfcU2fVersionCommand)]).get() reply:makeBlockPtr([this](NSData * _Nullable versionData, NSError * _Nullable error) {
-                if (compareCcidVersion(versionData, kCtapNfcAppletSelectionU2f, sizeof(kCtapNfcAppletSelectionU2f))) {
-                    callOnMainRunLoop([this] () mutable {
-                        if (m_service)
-                            m_service->didConnectTag();
-                    });
-                    return;
-                }
-            }).get()];
-    }).get()];
+        weakThis->transact(Vector { kCtapNfcAppletSelectionCommand, sizeof(kCtapNfcAppletSelectionCommand) }, [weakThis = WTFMove(weakThis)] (Vector<uint8_t>&& response) mutable {
+            ASSERT(RunLoop::isMain());
+            if (!weakThis)
+                return;
+            if (response == Vector { kCtapNfcAppletSelectionU2f, sizeof(kCtapNfcAppletSelectionU2f) }) {
+                if (weakThis && weakThis->m_service)
+                    weakThis->m_service->didConnectTag();
+            }
+        });
+    });
 }
 
 void CcidConnection::transact(Vector<uint8_t>&& data, DataReceivedCallback&& callback) const
 {
-    [m_smartCard transmitRequest:adoptNS([[NSData alloc] initWithBytes:data.data() length:data.size()]).autorelease() reply:makeBlockPtr([this, callback = WTFMove(callback)](NSData * _Nullable nsResponse, NSError * _Nullable error) mutable {
-        auto response = vectorFromNSData(nsResponse);
-        callOnMainRunLoop([this, response = WTFMove(response), callback = WTFMove(callback)] () mutable {
-            callback(WTFMove(response));
-            (void)this;
-        });
+    [m_smartCard beginSessionWithReply:makeBlockPtr([this, data = WTFMove(data), callback = WTFMove(callback)] (BOOL success, NSError *error) mutable {
+        if (success) {
+            [m_smartCard transmitRequest:adoptNS([[NSData alloc] initWithBytes:data.data() length:data.size()]).autorelease() reply:makeBlockPtr([this, callback = WTFMove(callback)](NSData * _Nullable nsResponse, NSError * _Nullable error) mutable {
+                auto response = vectorFromNSData(nsResponse);
+                callOnMainRunLoop([this, response = WTFMove(response), callback = WTFMove(callback)] () mutable {
+                    callback(WTFMove(response));
+                    (void)this;
+                });
+                [m_smartCard endSession];
```

Comment:
Is it possible for `m_smartCard` to be deallocated/invalidated through the `callOnMainRunloop() `code? Is it correct that we might (will?) call endSession before the code in `callOnMainRunLoop` runs?

Reply:
I've rearranged this, ending the session before calling callOnMainRunloop, to make this more clear. We don't need it after we have the response.